subsystems/networking/bluetooth.go

```diff
@@ -0,0 +1,698 @@
+				appAddress = appAddressInput
+			}
+		}
+
+		// If we've received all required credentials, break to pass them through inputChan.
+		if requiresWiFiCredentials && requiresCloudCredentials && //nolint:gocritic
+			ssid != "" && psk != "" && robotPartKeyID != "" && robotPartKey != "" && appAddress != "" {
+			break
+		} else if requiresWiFiCredentials && !requiresCloudCredentials && ssid != "" && psk != "" {
+			break
+		} else if requiresCloudCredentials && !requiresWiFiCredentials && robotPartKeyID != "" && robotPartKey != "" && appAddress != "" {
+			break
+		}
+
+		// If we haven't received all required credentials, try again a second later.
+		if !bsl.listenForCredentialsHealth.Sleep(ctx, time.Second) {
+			return nil // Return nil instead of a canceled context error to gracefully exit.
+		}
+		continue
+	}
+	inputChan <- userInput{SSID: ssid, PSK: psk, PartID: robotPartKeyID, Secret: robotPartKey, AppAddr: appAddress}
+	return nil
+}
+
+// listenForPairRequest ensures this device automatically accepts bluetooth pairing requests.
+func listenForPairRequest(ctx context.Context, bsl *bluetoothServiceLinux) error {
+	conn, err := dbus.SystemBus()
+	if err != nil {
+		return fmt.Errorf("failed to connect to system DBus: %w", err)
+	}
+	defer conn.Close() //nolint:errcheck
+
+	// Register bluez agent if not registered.
+	if bsl.bluezAgentRegistered {
+		bsl.logger.Debug("Bluez agent is already registered.")
+		return nil
+	}
+	reply := conn.Export(nil, BluezAgentPath, BluezAgent)
+	if reply != nil {
+		return fmt.Errorf("failed to export Bluez agent: %w", reply)
+	}
+	obj := conn.Object(BluezDBusService, "/org/bluez")
+	call := obj.Call("org.bluez.AgentManager1.RegisterAgent", 0, dbus.ObjectPath(BluezAgentPath), "NoInputNoOutput")
+	if err := call.Err; err != nil {
+		return fmt.Errorf("failed to register Bluez agent: %w", err)
+	}
+	call = obj.Call("org.bluez.AgentManager1.RequestDefaultAgent", 0, dbus.ObjectPath(BluezAgentPath))
+	if err := call.Err; err != nil {
+		return fmt.Errorf("failed to set default Bluez agent: %w", err)
+	}
+	bsl.logger.Debug("Bluez agent registered!")
+	bsl.bluezAgentRegistered = true
+
+	// Begin listening for property changes (i.e. bluetooth pairing requests)
+	if bsl.listeningForPropertyChanges {
+		bsl.logger.Debug("Already listening for property changes (bluetooth pairing requests) on the system D-bus.")
+		return nil
+	}
+	matchRule := "type='signal',interface='org.freedesktop.DBus.Properties',member='PropertiesChanged'"
+	err = conn.BusObject().Call("org.freedesktop.DBus.AddMatch", 0, matchRule).Err
+	if err != nil {
+		return fmt.Errorf("failed to add DBus match rule: %w", err)
+	}
+	bsl.listeningForPropertyChanges = true
+	bsl.logger.Debug("Listening for property changes (bluetooth pairing requests) on system D-bus.")
+
+	// Listen for properties changed events (includes bluetooth pairing requests).
+	signalChan := make(chan *dbus.Signal, 25)
+	conn.Signal(signalChan)
```

Comment:
Low priority: You're wrapping this entire function itself in a background routine. In theory, this function should do most of this inline (so these errors can get returned quickly during this "setup" phase) and then only background the loop itself below that will actually be waiting for stuff to happen.